cmd/tealdbg/debugger.go

```diff
@@ -89,19 +91,54 @@ type programMeta struct {
 	states       AppState
 }
 
-// breakpointLine is a source line number with a couple special values:
-// -1 do not break
-//  0 break at next instruction
-//  N break at line N
-type breakpointLine int
+// debugConfig contains information about control execution and breakpoints.
+type debugConfig struct {
+	NoBreak     bool `json:"nobreak"`
+	StepBreak   bool `json:"stepbreak"`
+	StepOutOver bool `json:"stepover"`
 
-const (
-	noBreak   breakpointLine = -1
-	stepBreak breakpointLine = 0
-)
+	ActiveBreak map[int]struct{} `json:"activebreak"`
+	CallDepth   int              `json:"calldepth"`
+}
 
-type debugConfig struct {
-	BreakAtLine breakpointLine `json:"breakatline"`
+func makeDebugConfig() debugConfig {
+	dc := debugConfig{}
+	dc.ActiveBreak = make(map[int]struct{})
+	return dc
+}
+
+func (dc *debugConfig) setNoBreak() {
+	dc.NoBreak = true
+}
+
+func (dc *debugConfig) setStepBreak() {
+	dc.StepBreak = true
+}
+
+func (dc *debugConfig) setStepOutOver(callDepth int) {
+	dc.StepOutOver = true
+	dc.CallDepth = callDepth
+}
+
+// setActiveBreak does not check if the line is a valid value, so it should
+// be called inside the setBreakpoint() in session.
+func (dc *debugConfig) setActiveBreak(line int) {
+	dc.ActiveBreak[line] = struct{}{}
```

Comment:
Andrew, I might have already asked but can't find the answer - what is the point of having `ActiveBreak` map? It is only set in `setActiveBreak`. And everywhere we call `setBreackpoint` the contol state is reset `s.debugConfig = makeDebugConfig()` leaving only a single entry in `ActiveBreak`.

Reply:
It's only used in the `Resume()` to set multiple active breakpoints (in case we jump backwards), but you're right that in everywhere else there is only 1 active break.